### PR TITLE
Add a concept of storage and appender as well as a basic main binary

### PR DESCRIPTION
**Pull request by dinowernli**
* Storage is currently backed by a single directory on disk
* The appender handles append requests until its log chunk is full, then writes it out
* Main just write a bunch of log entries

Sample run:

```
go run main.go
2017/09/29 23:54:04 Creating new storage with root: /tmp/yarder-dev482367775
2017/09/29 23:54:04 Wrote chunk file: /tmp/yarder-dev482367775/chunk-0
2017/09/29 23:54:04 Wrote chunk file: /tmp/yarder-dev482367775/chunk-1
2017/09/29 23:54:04 Wrote chunk file: /tmp/yarder-dev482367775/chunk-2
```

### CODE REVIEW

**Comment by dinowernli** [comment]
@devnev

File: src/github.com/log-yarder/yarder/storage/storage.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// Storage represents a single machine's interface to a corpus of almanac data.
+type Storage interface {
+	// CreateChunk allocates a new empty chunk of logs.
+	CreateChunk() (LogChunk, error)
+}
+
+// NewDiskStorage returns an instance of Strage backed by files on disk.
+func NewDiskStorage(rootPath string) Storage {
```
Comment (devnev) [inline src/github.com/log-yarder/yarder/storage/storage.go]: This constructor (and the appender one too) are just stuffing the values into matching fields and initialising the rest to zero. Seems like the struct could constructed directly instead.

Comment (dinowernli) [reply src/github.com/log-yarder/yarder/storage/storage.go]: Done.

File: storage/storage.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+func writeJson(path string, content interface{}) error {
+	bytes, err := json.Marshal(content)
+	if err != nil {
+		return fmt.Errorf("Unable to marshal json: %v", err)
+	}
+
+	err = ioutil.WriteFile(path, bytes, filePerms)
+	if err != nil {
+		return fmt.Errorf("Unable to write file to file %s: %v", path, err)
```
Comment (devnev) [inline storage/storage.go]: `write file to file`? did you mean `write to file`?

File: main/main.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
+package main
```
Comment (devnev) [inline main/main.go]: The binary is named after the directory containing the main package. Why not leave this at top-level for now?

Comment (dinowernli) [reply main/main.go]: Ah, didn't know you could have a package "main" in a directory not called "main". Fixed.

File: storage/storage.go (hunk 2 above)
Comment (dinowernli) [reply storage/storage.go]: Yup, typo. Fixed.

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,2 @@
+*swp
+main/main
```
Comment (devnev) [inline .gitignore]: update gitignore

File: main.go [hunk 5]
```diff
@@ -1,4 +1,30 @@
...
+	"io/ioutil"
+	"path"
+)
+
+const (
+	maxEntriesPerChunk = 10
+)
+
 func main() {
+	tmpDir, err := ioutil.TempDir(path.Join("/", "tmp"), "yarder-dev")
```
Comment (devnev) [inline main.go]: `ioutil.TempDir`:
> If dir is the empty string, TempDir uses the default directory for temporary files (see os.TempDir)

File: main.go [hunk 6]
```diff
@@ -1,4 +1,30 @@
...
+	if err != nil {
+		panic(fmt.Sprintf("Unable to create temp dir, %v", err))
```
Comment (devnev) [inline main.go]: Use the `log` package rather than panicking.

File: storage/storage.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path"
+)
+
+const (
+	dirPerms  = 0777
+	filePerms = 777
```
Comment (devnev) [inline storage/storage.go]: octal fileperms, don't make the files executable -> `0644` is usually fine.

File: storage/storage.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path"
+)
+
+const (
+	dirPerms  = 0777
```
Comment (devnev) [inline storage/storage.go]: world-writable by default probably doesn't make sense. `0755`?

**Review by devnev — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 4 above)
Comment (dinowernli) [reply .gitignore]: Done

File: main.go (hunk 5 above)
Comment (dinowernli) [reply main.go]: Cool, done.

File: main.go (hunk 6 above)
Comment (dinowernli) [reply main.go]: Done.

File: storage/storage.go (hunk 7 above)
Comment (dinowernli) [reply storage/storage.go]: Done.

File: storage/storage.go (hunk 8 above)
Comment (dinowernli) [reply storage/storage.go]: Done.

**Comment by dinowernli** [comment]
PTAL @devnev

File: main.go [hunk 9]
```diff
@@ -1,4 +1,31 @@
...
+	"log"
+	"os"
+)
+
+const (
+	maxEntriesPerChunk = 10
+)
+
 func main() {
+	tmpDir, err := ioutil.TempDir(os.TempDir(), "yarder-dev")
```
Comment (devnev) [inline main.go]: you don't need to call `os.TempDir` at all, just pass `""` to `ioutil.TempDir`

File: .gitignore [hunk 10]
```diff
@@ -0,0 +1,2 @@
+*swp
+main
```
Comment (devnev) [inline .gitignore]: the binary will be called `yarder`, not `main`

**Review by devnev — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 10 above)
Comment (dinowernli) [reply .gitignore]: Done.

File: main.go (hunk 9 above)
Comment (dinowernli) [reply main.go]: Done.

**Comment by dinowernli** [comment]
Thanks, PTAL

**Review by devnev — APPROVED** [review]

**Pull request merged**